examples/stencils.cpp

```diff
@@ -0,0 +1,237 @@
+
+// Helper output
+static std::ostream &labeled_cout() {
+  return std::cout << "Thread: " << std::hex << std::this_thread::get_id()
+                   << " ";
+}
+
+// Cache of compiled scripts
+class JSCache {
+ public:
+  ~JSCache() {
+    labeled_cout() << "Destructing cache with " << _cache.size()
+                   << " scripts\n";
+  }
+
+  RefPtr<JS::Stencil> find(std::string const &key) {
+    std::lock_guard<std::mutex> lk(_mx);
+    auto it = _cache.find(key);
+    return it == _cache.end() ? nullptr : it->second;
+  }
+
+  void insert(std::string const &key, RefPtr<JS::Stencil> const &val) {
+    std::lock_guard<std::mutex> lk(_mx);
+    _cache.insert(std::pair{key, val});
+  }
+
+ private:
+  std::mutex _mx;
+  std::map<std::string, RefPtr<JS::Stencil>> _cache;
+};
+
+// Script compiler and executer
+class Job {
+ public:
+  Job(JSCache &cache);
+  ~Job() {
+    if (_fc) JS::DestroyFrontendContext(_fc);
+  };
+
+  void execute_script(JSContext *cx, std::string const &script,
+                      const char *filename, unsigned long linenumber);
+
+ private:
+  RefPtr<JS::Stencil> compile_script(JSContext *cx, std::string const &script,
+                                     const char *filename,
+                                     unsigned long linenumber);
+
+ private:
+  JSCache &_cache;
+  JS::FrontendContext *_fc;
+};
+
+Job::Job(JSCache &cache) : _cache(cache) {
+  static constexpr size_t kCompileStackQuota = 128 * sizeof(size_t) * 1024;
+  _fc = JS::NewFrontendContext();
+  if (_fc) JS::SetNativeStackQuota(_fc, kCompileStackQuota);
+}
+
+void Job::execute_script(JSContext *cx, std::string const &script,
+                         const char *filename, unsigned long linenumber) {
+  JS::RootedScript rscript(cx);
+  RefPtr<JS::Stencil> stencil(_cache.find(script));
+
+  if (stencil == nullptr) {
+    labeled_cout() << "Compiling script\n";
+    stencil = compile_script(cx, script, filename, linenumber);
+    if (stencil != nullptr) _cache.insert(script, stencil);
+  } else {
+    labeled_cout() << "Taking script from the cache\n";
+  }
+
+  if (stencil == nullptr) {
+    boilerplate::ReportAndClearException(cx);
+    return;
+  }
+
+  JS::InstantiateOptions instantiateOptions;
+  rscript = JS::InstantiateGlobalStencil(cx, instantiateOptions, stencil);
+  if (!rscript) {
+    boilerplate::ReportAndClearException(cx);
+    return;
+  }
+
+  JS::RootedValue val(cx);
+  if (!JS_ExecuteScript(cx, rscript, &val)) {
+    boilerplate::ReportAndClearException(cx);
+    return;
+  }
+}
+
+RefPtr<JS::Stencil> Job::compile_script(JSContext *cx,
+                                        std::string const &script,
+                                        const char *filename,
+                                        unsigned long linenumber) {
+  JS::SourceText<mozilla::Utf8Unit> source;
+  if (!_fc) return nullptr;
+
+  JS::CompileOptions opts(cx);
+  opts.setFileAndLine(filename, linenumber);
+  opts.setNonSyntacticScope(true);
```

Comment:
This `setNonSyntacticScope` call looks unnecessary.
Non-syntactic scope flag is necessary when you want to pass environment chain when executing.
the flag affects how the non-local variable works, and if it's true, the non-local variable gets slightly de-optimized, in order to access the variable in the environment chain.

Reply:
My other code crashed without the call, and I left it here because I didn't understand what it did. I removed the call.